addons/im_livechat/models/chatbot_script_step.py

```diff
@@ -0,0 +1,257 @@
+    step_type = fields.Selection([
+        ('text', 'Text'),
+        ('question_selection', 'Question'),
+        ('question_email', 'Email'),
+        ('question_phone', 'Phone'),
+        ('forward_operator', 'Forward to Operator'),
+        ('free_input_single', 'Free Input'),
+        ('free_input_multi', 'Free Input (Multi-Line)'),
+    ], default='text', required=True)
+    # answers
+    answer_ids = fields.One2many(
+        'chatbot.script.answer', 'script_step_id',
+        copy=True, string='Answers')
+    triggering_answer_ids = fields.Many2many(
+        'chatbot.script.answer', domain="[('script_step_id.sequence', '<', sequence)]",
+        compute='_compute_triggering_answer_ids', readonly=False, store=True,
+        copy=False,  # copied manually, see chatbot.script#copy
+        string='Only If', help='Show this step only if all of these answers have been selected.')
+    # forward-operator specifics
+    is_forward_operator_child = fields.Boolean(compute='_compute_is_forward_operator_child')
+
+    @api.depends('sequence')
+    def _compute_triggering_answer_ids(self):
+        for step in self:
+            if step.triggering_answer_ids:
+                update_command = [Command.unlink(answer.id) for answer in step.triggering_answer_ids
+                                  if answer.script_step_id.sequence >= step.sequence]
+                if update_command:
+                    step.triggering_answer_ids = update_command
+
+    def _get_parent(self):
+        """ Returns the first preceding step that matches either the triggering answers
+         or the possible answers the user can select """
+        self.ensure_one()
+
+        if not self.chatbot_script_id.ids:
+            return False
+
+        domain = [
+            ('chatbot_script_id', 'in', self.chatbot_script_id.ids),  # Used ids instead of id to avoid testing on NewId
+            ('step_type', 'in', ['forward_operator', 'question_selection']),
+            ('sequence', '<', self.sequence),
+        ]
+        if self.triggering_answer_ids:
+            answer_domain = []
+            for answer in self.triggering_answer_ids:
+                answer_domain = expression.AND([answer_domain, [('triggering_answer_ids', 'in', answer.ids)]])
+            domain = expression.AND([domain, expression.OR([
+                answer_domain, [('answer_ids', 'in', self.triggering_answer_ids.ids)]])
+            ])
+        else:
+            domain = expression.AND([domain, [('triggering_answer_ids', '=', False)]])
+        return self.search(domain, order='sequence desc', limit=1)
+
+    @api.depends('sequence', 'triggering_answer_ids')
+    def _compute_is_forward_operator_child(self):
+        for step in self:
+            parent = step._get_parent()
+            while parent:
+                if parent.step_type == 'forward_operator':
+                    break
+                parent = parent._get_parent()
```

Comment:
To remove a bit of redundancy

```suggestion
            while True:
                parent = parent._get_parent()
                if parent.step_type == 'forward_operator':
                    break
```

isn't a better way to do it ?

Here you will do a search in 2 nested loops which is very inefficient... 

What can be done e.g. is to fetch all `script.step` at the beginning of the compute, and then do `filtered_domain` to find the parent

In `_get_parent`, give a list of script.step,
- if not given, it will perform the search like it does now
- if a list of script.step is given, it will do `filtered_domain` on them

What do you think ?